### DISCUSSION TITLE
Cannot read properties of null (reading 'setItem')

### DISCUSSION THREAD

**Discussion by bhoodream** [pmndrs/zustand#1799] [General] [answered]
## Summary
```
TypeError
Cannot read properties of null (reading 'setItem')
```

Found this error in sentry. The reason is clearly that the `localStorage` at the time `setItem` is called is `null`. The error occurs quite rarely.

Unfortunately, I don't have time to try to reproduce it.

<img width="1143" alt="Снимок экрана 2023-05-11 в 12 45 08" src="https://github.com/pmndrs/zustand/assets/1740470/729b8250-a7bb-40ad-ac4a-f9d82362a57a">

**Comment by dai-shi** [comment]
Hope anyone can reproduce it.

**Comment by bhoodream** [comment]
All devices with an error are something like this

<img width="1115" alt="Снимок экрана 2023-05-11 в 13 56 35" src="https://github.com/pmndrs/zustand/assets/1740470/879321dc-ff0d-4292-adda-682fde6739f9">

**Reply by dbritto-dev** [reply]
@bhoodream would you mind sharing us a basic example of your code on codesandbox?

**Reply by bhoodream** [reply]
The application is not the smallest one, I think it will not be easy to reproduce all the nuances in the example.

In fact, this is a fairly standard React application built using webpack. It does not seem that this problem is due to the peculiarities of our application... It is also not at all clear why the error is so rare and occurs on typical devices with the Android 13 system.

The most likely reason is some kind of restriction of access rights to localStorage on these systems. It seems to me that your library should try to handle such a case, when the localStorage can be null.

**Reply by dbritto-dev** [reply]
@dai-shi @bhoodream seems some WebView Browsers doesn't support `window.localStorage` so it can be `null` due to that or due to not enough permissions. So in that case what should be the behavior?

**Reply by bhoodream** [reply]
As a user of the persist middleware, I would like my application to keep running, even without the ability to persist state for some reason. 

For me, the application crash due to the fact that the `localStorage` is not available is not the most predictable and expected behavior.

**Reply by dai-shi** [reply]
Meanwhile, you could create your own storage without using createJSONStorage to fit with your use case. createJSONStorage isn't designed to cover all cases.

**Reply by bhoodream** [reply]
In the current implementation, you are ready for the fact that it may not be possible to get a store.
```
 try {
    storage = getStorage()
  } catch (e) {
    // prevent error if the storage is not defined (e.g. when server side rendering a page)
    return
  }
```

Why not provide for a case where the definition of the store might be there, but the store itself might not be?

In addition, you use `as`, thereby acting on faith, is it worth it to play it safe with the help of a `try / catch` construction?
```
// your implementation
setItem: (name, newValue) =>
      (storage as StateStorage).setItem(
        name,
        JSON.stringify(newValue, options?.replacer)
      ),

// safe implementation
setItem: (name, newValue) => {
try {
     (storage as StateStorage).setItem(
        name,
        JSON.stringify(newValue, options?.replacer)
      )
  } catch (e) {
    // prevent error if the storage is not exists
  }
}
```

**Reply by dai-shi** [reply]
try/catch is bad because it hides other errors. Someone reported such issue in the past.
Do you mean `(storage as StateStorage | undefined)?.setItem(` solves your issue?

**Reply by bhoodream** [reply]
`try / catch` was just an example, you know better how to do it.

Yes, the optional method call solves my problem in theory.

**Reply by dai-shi** [reply]
Because I don't follow the discussion fully, I'm not sure how it solves.
Please give it a try and see if the optional method call really solves your problem.

Maybe @dbritto-dev can help?

**Reply by dbritto-dev** [reply]
In fact we can do something like this:

```ts
const memoryStorage = {
  setItem(name, value) {
    this[`_${name}`] = value
  } ,
  getItem(name) { 
    return this[`_${name}`]
  },
  removeItem(name) {
    this[`_${name}`] = undefined
  },
}

const localStorageSupported = () => {
  try { 
    const x = "__storage_test__"
    localStorage.setItem(x, x)
    localStorage.removeItem(x)
    return true
  } catch {
    console.warning(`localStorage is unsupported or unavailable`)
    return false
  }
}
```

```ts
storage: createJSONStorage(() => {
  if (localStorageSupported()) return localStorage
  return memoryStorage
})
```

**Reply by bhoodream** [reply]
Yes, that's what I was talking about. When using the default implementation, you do not expect that you will still have to handle some edge cases. With such a check, the code will work, even if the `localStorage` is unavailable for some reason.

And thanks to the warning in the console, users of this middleware will be able to understand what is wrong.

**Reply by dbritto-dev** [reply]
@bhoodream would youd mind using that solution for now?

@dai-shi any thougts? I can add it to persist middleware or add it to the persist middleware guide. In both cases people can take into account what could happen.

**Reply by dai-shi** [reply, marked as answer]
@dbritto-dev I think that is usage issue, so improving docs should help.

This pattern should be something we can recommend in docs.
```js
const myStorage = {
  getItem(name) {
    const value = JSON.parse(localStorage.getItem(name));
    // preferably validate value
    return value;
  },
  setItem(name, value) {
    try {
      localStorage.setItem(name, JSON.stringify(value));
    } catch (e) {
      console.warn(...);
    }
  },
  removeItem(name) {
    localStorage.removeItem(name);
  },
};

export const useBearStore = create(
  persist(
    (set, get) => ({
      bears: 0,
      inc: () => set((state) => ({ bears: state.bears + 1 })),
    }),
    {
      name: 'bear-storage',
      storage: myStorage
    }
  )
)
```

**Reply by bhoodream** [reply]
Yes, now I agree that this problem should be solved on the side of the middleware user. If you focus on this a little in the documentation, I think there will be no such misunderstanding. And if it does, then you won’t have to spend a lot of time explaining it, but immediately point to the recommendations.

Thanks for the quick help :)

**Reply by dbritto-dev** [reply]
Cool, I'll add it to docs

**Reply by dai-shi** [reply]
> Cool, I'll add it to docs

@dbritto-dev Sounds great.

> Meanwhile, you could create your own storage without using createJSONStorage to fit with your use case. createJSONStorage isn't designed to cover all cases.

Note again that createJSONStorage isn't ideal, as it relies on JSON.parse which isn't type safe, and there's no validations.
We should consider that createJSONStorage is a _fallback_ mechanism. Preferably, people are encouraged to create their own storage for their needs with type safety and validations.

**Reply by dbritto-dev** [reply]
> > Cool, I'll add it to docs
> 
> @dbritto-dev Sounds great.
> 
> > Meanwhile, you could create your own storage without using createJSONStorage to fit with your use case. createJSONStorage isn't designed to cover all cases.
> 
> Note again that createJSONStorage isn't ideal, as it relies on JSON.parse which isn't type safe, and there's no validations.
> We should consider that createJSONStorage is a _fallback_ mechanism. Preferably, people are encouraged to create their own storage for their needs with type safety and validations.

@dai-shi yeah, that makes sense.

**Reply by zupat** [reply]
@dai-shi  `getItem` can also fail if cookies are disabled.

**Reply by dai-shi** [reply]
Maybe wrapping with try-catch should help.
